### DISCUSSION TITLE
How to handle CORS and Same Origin Policy issues?

### DISCUSSION THREAD

**Discussion by Kukulkano** [webui-dev/webui#157] [General]
Hi. I plan to write some client software and found WebUI as a great option. But the web application I want to run uses WASM and is making calls to some websites to get data (API REST calls).

In the past, it always failed because of missing sufficient CORS headers on server side and issues with the Same Origin Policy. I understand that the API holder does not want to send a CORS header with *.  If I want to run a web app in WebUI (or any other web gadget) I always run against these issues. I would need a way to force the used webbrowser to ignore CORS and Same Origin Policy. But I found no such option. Is there a practical way to get around these limitations? How do you guys handle these issues?

**Comment by AlbertShown** [comment]
The first thing to clarify is that WebUI is only a tool to connect your application to the web browser to use it as your main GUI. All other things, like CORS, are entirely out of control by WebUI.

Normally, the server-side control handles the request or ignores it. So, the client should respect the server-side rules and put the necessary headers. Keep in mind that your application is running on `127.0.0.1`. So your client can easily modify the header before sending it using `xhr.setRequestHeader('custom-header', 'value');`.

**Reply by Kukulkano** [reply]
Thanks Albert. I know that WebUI is out of control for that. This is why I wondered how other WebUI users are handling these issues. I tried to write a web app several times before and always failed on Same Origin Policy (my app from 127.0.0.1 trying to access foreign website) and CORS (server did not set "Access-Control-Allow-Origin: ".

I assumed that maybe people are piping all network IO through the function calls (WebUI -> JS -> WebUI) and then using CURL or whatever that is not caring about Same Origin Policy or CORS. Or maybe more easy by some sort of Proxy implementation?

Anyway, with WebUI it still is hard to do an application that needs to access Web-APIs using REST calls. So sad.

Anyway, WebUI is a great tool and I like it very much. Maybe I find a way to use this in the future.

**Reply by delphius** [reply]
I'll just leave it here 😁
https://github.com/delphius/fsclient/blob/58fdc58df56b446ed56e6d67dd0e8a6cc26983dc/xpath.js#L15

**Reply by Kukulkano** [reply]
Hi Delphius. Thanks, but how does the linked xpath module help me with my question? Did I miss an important point?

**Reply by AlbertShown** [reply]
Your application is written in what language?

**Reply by Kukulkano** [reply]
Hi Albert. It is not written yet. I plan to use golang. I'm just doing some pre-tests and evaluation. This includes testing some functionality in a webbrowser first. Using a proxy to work around the CORS headers like suggested by delphius is working, but not yet a final solution as I cannot rely on such third party proxy service for a commercial app. But I think I can solve this.

More sadly, I still fail when loading a WASM webworker in Chrome from local files (Failed to construct 'Worker'). It seems very hard to develop anything that comes from local filesystem and needs some more than simple functionality...

**Reply by AlbertShown** [reply]
We are then on the same path 🐱‍👤

I work in a company, and we have two major commercial projects, one in C++ and a new one in Golang. My team spend time adding new functionality in C++ GUI, then more time to add the same things in the Golang GUI project... wish was a total waste of time.

A while ago, I suggested webui to my supervisor, and the next day, I became the hero of my team!
Now, we are using webui 2.0.3 in both projects, using web technologies makes our UI looks better and more modern and professional. On top of that we are using the same UI module (HTML/JS/CSS) in both projects, which makes us save a lot of time. And the happy ending, I'm paid for my contribution to webui as I'm allowed to do it during the working hour. Two more engineers in my team will contribute soon to webui.

In our Golang project, we also communicate to a web server through HTTP APIs, but it's our web server, so we don't have the CORS issue as we allow `origin *`. But I can suggest two or three options for your project. Did your final product has a setup package, or it's a single executable binary file?

**Reply by Kukulkano** [reply]
Hi Albert. Very interesting! Thanks for sharing your experience.

Today we have client software for Windows, Linux, MacOS, iOS and Android (end user software). You may imagine what a pain this is for development... But we need to react on file types registered on the client OS. We need to run our app if a specific file type is opened (double clicked). In order to get this, we will use a setup for sure. This will register the file extensions and/or mime types. If executed, it will hand over the file content (eg as byteArray) to the loaded website (in WebUI if possible). This is what I'm currently evaluating.

Sadly, it looks like providing the source as local files is opening up to many issues (same origin, CORS, WebWorker issues etc). So I just consider providing the whole source by a https website. This seems to work better than local files. So if the app starts, the WebUI just pushes a https URL to get all app code from some online URL. If the used web browser uses caching, it should load fast enough :-)

I have not solved all open questions and all open issues but for the moment it looks great (still many questions related to other things than WebUI). Not sure what happens in companies with proxy servers (Auth, PAC files etc). All things to consider and think about...

**Reply by delphius** [reply]
My decision was due to the need to circumvent this limitation in proof of concept mode. In production mode, the two main options are to raise your cors proxy server, for example https://github.com/Rob--W/cors-anywhere or rely on a proven service that can be used anyway https://developers.cloudflare.com/workers/examples/cors-header-proxy/

**Reply by Kukulkano** [reply]
Hi Albert. May I ask you about your experience with WebUI and end users? I mean, do you ask people to install a specific web browser? Or do you rely on WebUI to find a suitable one? If it does not find one, do you ask people to install a specific one? Do you have experience in reliability of such solution? Do many people have to install additional web browser? I mean, thousands of our end users are not IT experts. More like the average WhatsApp user. They would not be able or allowed to install other software...

**Reply by AlbertShown** [reply]
- We rely on webui to figure out the web browser. In most cases, it's Chrome. Otherwise, it's the currently running web browser
- Our product (Software) comes as a setup package, so all needed files are already included
- We decided to go with the DLL version, so our installer can easily update the webui dll if needed.
- No issues are reported, except one user (our support guy) said he close Yandex first to force webui to pick up Chrome (So, maybe it's a good idea if we add `webui_disable_specific_browser(Yandex);`)
- In general, the end-user does even notice he is using a web browser... they even congrats us because they like the new UI design
- Our software controls a big machine through LAN, so all that webui does is send clicks events and update the motors speed value through JavaScript in  the UI... that's all, so that's why it's working perfectly for us for now
- The last thing... consider using something other than the current repo version. It's a mess right now!, clone 2.0.3 and patch it as needed instead

**Reply by Kukulkano** [reply]
Thanks Albert. Sounds interesting. Our app will have much more functionality. It has to integrate with the OS (to handle specific files) and it has to do web API calls and to run a WASM component (encryption and compression stuff). The GUI itself is the minor part.

So you don't ship a web browser, right? You just rely on WebUI finding a usable and up to date browser. I wonder what happens if we do so for >10.000 end users... Is someone having expericence with WebUI in that field (many thousand end users)? If that would trigger a support call for every 20th user, it would be a nightmare.

For the repo version, this was a good hint. I also had a look and found that WebUI is not maintaining an official stable or LTS branch (for example for only fixing bugs but not implementing any new features). We've had very bad experiences with such components in the past. Every new version we were forced to pull for fixing a bug also introduced new bugs because of new functions and API changes. By this, we were forced to do a full QA cycle for every time we did a pull on that component for just getting rid of an annoying bug. That was very bad and cost us many many hours and also some hair... This is definitely a disadvantage of WebUI :-(

**Reply by AlbertShown** [reply]
>> WebUI is not maintaining an official stable or LTS branch

In our case, we have our own private webui 2.0.3 LTS. I submitted to this current branch most patches we did locally.

>> Is someone having expericence with WebUI

Our project is used right now by our customers only (_companies_), so around 300 - 500 technicians around the globe. Our UI is simple, that's why our QA validate it in the first test.

>> I wonder what happens if we do so for >10.000 end users

In our case, the UI does not have any open/read files, but when I tested the text-editor example, I noticed that sometimes the OpenFile dialog in Windows loses the focus (_Shows behind the UI window_), not a big deal but QA may report a bad UX. But this issue never happens in Linux or macOS.

>> So you don't ship a web browser, right?

If we need to ship a web browser, then we will switch to React and re-write the whole project in JavaScript. The main reason for using webui is to make our software package very small, multi-platform, modern UI... and the most unique and fantastic feature of webui is we can use our favourite programming language in the backend.

>> has to do web API calls and to run a WASM component

I have no WASM experience, but someone reported that WASM won't run correctly in any local host. So, you may need to patch webui to change the domain or something... I'm not sure.

**Reply by Kukulkano** [reply]
Thanks Albert. 

> someone reported that WASM won't run correctly in any local host

IMHO, the main issue is, that WebWorkers do not work at all if the files are loaded from localhost :-( But that is not a WebUI issue.

> sometimes the OpenFile dialog in Windows loses the focus (Shows behind the UI window)

This is weird but maybe not a real issue. Maybe even a problem with only a specific browser? But thanks for pointing to it.

> In our case, we have our own private webui 2.0.3 LTS. I submitted to this current branch most patches we did locally.

So, basically, you say you run your own stable local branch and cherry pick any bug fixes from the master branch into it? I understand. But from my experience, cherry picking patches can have bad side effects if you're not deep in the project code by itself. As I do not plan to dig deeper into WebUI itself, this is possibly a dangerous plan for my project. I will have to think about this.

**Comment by delphius** [comment]
> Hi Delphius. Thanks, but how does the linked xpath module help me with my question? Did I miss an important point?

Hi, this is not just an xpath module, it is part of a general program using webui, which parses data from external sources to form a representation. As already mentioned, webui by itself is not able to solve the CORS problem in any way, and the presented module is a solution to your problem using a special publicly available free CORS proxy. You asked how webui users solve this problem, I showed you a specific way that works for me, without any complications.

For your purposes, only one specific line is enough, the link to which I gave, so that firstly there was a clear example of practical use, and secondly, on the site itself, which provides this specialized proxy service, you can get the details of use. But, I repeat, this has nothing to do with the webui library used at all, it is rather a practical issue that falls into its ecosystem when applied in practice.

Or maybe I completely misunderstood the question, sorry.

**Reply by Kukulkano** [reply]
Ah, thanks! I did not understand your intention, sorry. I will have a look!

**Reply by delphius** [reply]
No problem. Fell free to contact me for any question if any
